pystachio/choice.py

```diff
@@ -0,0 +1,166 @@
+        else:
+            return self._value
+
+    def value(self):
+        return self._value
+
+    def dup(self):
+        return self.__class__(self._value)
+
+    def __hash(self):
+        return hash(self.get())
+
+    def __repr__(self):
+        si, _ = self.interpolate()
+        return '%s(%s)' % (self.__class__.__name__,
+                           repr(self._value))
+
+    def __eq__(self, other):
+        if not isinstance(other, ChoiceContainer):
+            return False
+        if len(self.CHOICES) != len(other.CHOICES):
+            return False
+        for myalt, otheralt in zip(self.CHOICES, other.CHOICES):
+            if myalt.serialize_type() != other.serialize_type():
+                return False
+        si, _ = self.interpolate()
+        oi, _ = other.interpolate()
+        return si._value == other._value
+
+    def check(self):
+        # Try each of the options in sequence:
+        # There are three cases for matching depending on the value:
+        # (1) It's a pystachio value, and its type is the type alternative. Then typecheck
+        #    succeeds.
+        # (2) It's a pystachio value, but its type is not the current alternative. Then the
+        #    typecheck proceeds to the next alternative.
+        # (3) It's not a pystachio value. Then we try to coerce it to the type alternative.
+        #    If it succeeds, then the typecheck succeeds. Otherwise, it proceeds to the next
+        #    type alternative.
+        # If none of the type alternatives succeed, then the check fails. match
+        for opt in self.CHOICES:
+            if isinstance(self._value, opt):
+                return self._value.in_scope(*self.scopes()).check()
+            # Try to do a coercion.
+            else:
+                try:
+                    tc = opt(self._value).in_scope(*self.scopes()).check()
+                    if tc.ok():
+                        return tc
+                except (self.CoercionError, ValueError):
+                    pass
+        # If we've reached here, then it failed all of its choices.
+        return TypeCheck.failure(
+            "%s typecheck failed: value %s did not match any of its alternatives" %
+            (self.__class__.__name__, self._value))
+
+    def interpolate(self):
+        if isinstance(self._value, Object):
+            return self._value.in_scope(*self.scopes()).interpolate()
+        else:
+            # If the value isn't a Pystachio object, then it needs to get wrapped in order to
+            # interpolate it. But we don't know what type it's intended to have. So we need to try
+            # to wrap it in the various type alternatives. The _first_ one that succeeds is
+            # what we want to use.
+            for choice_type in self.CHOICES:
+                try:
+                    # If this succeeds, then return it. If not, try the next type.
+                    return choice_type(self._value).in_scope(*self.scopes()).interpolate()
```

Comment:
I don't think so - won't that lose the scopes defined on the choice object? (Tests seem to indicate that that's what's happening.)